numpy/core/src/multiarray/textreading/rows.c

```diff
@@ -246,6 +257,56 @@ read_rows(stream *s,
             // We've deferred some of the initialization tasks to here,
             // because we've now read the first line, and we definitively
             // know how many fields (i.e. columns) we will be processing.
+
+            if (usecols_obj != Py_None) {
+                // Call the Python function provided by the caller to
+                // create the usecols array.
+                PyObject *seq = PyObject_CallFunction(usecols_obj, "n",
+                                                      current_num_fields);
+                if (PyErr_Occurred()) {
+                    // User-provided function failed.
+                    goto error;
+                }
+                // The user-defined usecols function must return a sequence
+                // of integers.
+                num_usecols = PySequence_Length(seq);
+                if (num_usecols == -1) {
+                    // User-provided function did not return a sequence.
+                    PyErr_Clear();
+                    PyErr_Format(PyExc_TypeError,
+                        "the user-provided callable usecols must return a "
+                        "sequence of ints, but it returned an instance of "
+                        "type '%s'", Py_TYPE(seq)->tp_name);
+                    Py_DECREF(seq);
+                    goto error;
```

Comment:
I am wondering if the "length check" can't be just part of the helper?  We use the same error in any case?
Could even use `PySequence_Fast` in the helper, but that doesn't really matter.

Reply:
That would mean that the helper should probably return `num_usecols` and fill in `usecols` (passed by pointer).